### ISSUE TITLE
[BUG] The handshake Origin header doesn't respect the URI scheme

### ISSUE THREAD

**Issue by jvpernis** [websocket-client/websocket-client#787]
When performing the handshake while connecting to a server, the Origin header is added automatically.
As you can see in the following snippet, the header's value is hardcoded to the HTTP scheme:
https://github.com/websocket-client/websocket-client/blob/81fa70d259ce1609b3fb1bf66c8a9bed81a3f28c/websocket/_handshake.py#L89-L93

However, in our setup we are primarily using HTTPS traffic to serve the clients. As such, connections from non-HTTPS origins are blocked by our Flask-SocketIO server. We are therefore not able to setup a connection using the `websocket-client` package.

**Note:** I am aware that the `suppress_origin` option exists, but we cannot use it since we are using the `python-socketio` package which does not expose it.

Here is the sample code I used:
```python
import socketio

sio = socketio.Client(logger=True, engineio_logger=True)
sio.connect("wss://<domain>")
```

<details>
  <summary>Debug trace</summary>
  
  
```
[2022-02-05 01:40:07,210] _connect_websocket(): Attempting WebSocket connection to wss://<domain>/socket.io/?transport=websocket&EIO=4
--- request header ---
[2022-02-05 01:40:07,213] dump(): --- request header ---
GET /socket.io/?transport=websocket&EIO=4&t=1644021607.2104435 HTTP/1.1
Upgrade: websocket
Host: <domain>
Origin: http://<domain>
Sec-WebSocket-Key: 1v3UqNV7XtEhCrLuHcrpkQ==
Sec-WebSocket-Version: 13
Connection: Upgrade


[2022-02-05 01:40:07,213] dump(): GET /socket.io/?transport=websocket&EIO=4&t=1644021607.2104435 HTTP/1.1
Upgrade: websocket
Host: <domain>
Origin: http://<domain>
Sec-WebSocket-Key: 1v3UqNV7XtEhCrLuHcrpkQ==
Sec-WebSocket-Version: 13
Connection: Upgrade


-----------------------
[2022-02-05 01:40:07,213] dump(): -----------------------
--- response header ---
[2022-02-05 01:40:07,213] trace(): --- response header ---
HTTP/1.1 101 Switching Protocols
[2022-02-05 01:40:07,214] trace(): HTTP/1.1 101 Switching Protocols
Server: nginx
[2022-02-05 01:40:07,214] trace(): Server: nginx
Date: Sat, 05 Feb 2022 00:40:07 GMT
[2022-02-05 01:40:07,214] trace(): Date: Sat, 05 Feb 2022 00:40:07 GMT
Connection: upgrade
[2022-02-05 01:40:07,214] trace(): Connection: upgrade
Upgrade: websocket
[2022-02-05 01:40:07,214] trace(): Upgrade: websocket
Sec-WebSocket-Accept: nhyrKZvlfR1QqPnrdhvAM7gji6w=
[2022-02-05 01:40:07,214] trace(): Sec-WebSocket-Accept: nhyrKZvlfR1QqPnrdhvAM7gji6w=
Strict-Transport-Security: max-age=0;
[2022-02-05 01:40:07,214] trace(): Strict-Transport-Security: max-age=0;
Access-Control-Allow-Origin: http://<domain>
[2022-02-05 01:40:07,215] trace(): Access-Control-Allow-Origin: http://<domain>
Access-Control-Allow-Credentials: true
[2022-02-05 01:40:07,215] trace(): Access-Control-Allow-Credentials: true
-----------------------
[2022-02-05 01:40:07,215] trace(): -----------------------
++Rcv raw: b'\x88\x00'
[2022-02-05 01:40:07,215] trace(): ++Rcv raw: b'\x88\x00'
++Rcv decoded: fin=1 opcode=8 data=b''
[2022-02-05 01:40:07,215] trace(): ++Rcv decoded: fin=1 opcode=8 data=b''
++Sent raw: b'\x88\x82\x14{a>\x17\x93'
[2022-02-05 01:40:07,215] trace(): ++Sent raw: b'\x88\x82\x14{a>\x17\x93'
++Sent decoded: fin=1 opcode=8 data=b'\x03\xe8'
[2022-02-05 01:40:07,215] trace(): ++Sent decoded: fin=1 opcode=8 data=b'\x03\xe8'
Traceback (most recent call last):
  File "./websocketDaemon.py", line 372, in <module>
    main()
  File "./websocketDaemon.py", line 366, in main
    sio.connect("wss://<domain>")
  File "/home/user/.local/lib/python3.8/site-packages/socketio/client.py", line 331, in connect
    self.eio.connect(real_url, headers=real_headers,
  File "/home/user/.local/lib/python3.8/site-packages/engineio/client.py", line 185, in connect
    return getattr(self, '_connect_' + self.transports[0])(
  File "/home/user/.local/lib/python3.8/site-packages/engineio/client.py", line 473, in _connect_websocket
    open_packet = packet.Packet(encoded_packet=p)
  File "/home/user/.local/lib/python3.8/site-packages/engineio/packet.py", line 27, in __init__
    self.decode(encoded_packet)
  File "/home/user/.local/lib/python3.8/site-packages/engineio/packet.py", line 51, in decode
    b64 = not self.binary and encoded_packet[0] == 'b'
IndexError: string index out of range
```

</details>

The error occurs because the Flask-SocketIO server closed the connection due to the invalid `Origin` header.
It logged the following error message:

```
http://<domain> is not an accepted origin.
```

**Comment by engn33r** [comment]
I skimmed the issue thread on python-socketio briefly so maybe I am just confused, but websocket-client allows you to set the full "Origin" header value with whatever protocol or text you want. The http header is hard coded for one branch of the if statement that you highlight, but the other branch allows for custom values. If you are not able to access this if statement branch because you are not using websocket-client directly, the issue may not be in websocket-client. Here is an example of setting a https origin:

```
import websocket

websocket.enableTrace(True)
ws = websocket.WebSocket()
ws.connect("ws://echo.websocket.events/", origin="https://google.com")
ws.send("Hello, Server")
print(ws.recv())
ws.close()
```

You can even set an Origin header which doesn't make any sense in normal usage, but can be useful for testing purposes:

```
import websocket

websocket.enableTrace(True)
ws = websocket.WebSocket()
ws.connect("ws://echo.websocket.events/", origin="!?!?!")
ws.send("Hello, Server")
print(ws.recv())
ws.close()
```

**Comment by jvpernis** [comment]
Indeed, `websocket-client` allows you to either suppress or override the Origin header, but in this particular case, I would consider that as just a workaround. 


This bug is about the Origin header not reflecting whether TLS protection used. If a secure WebSocket connection is used, the Origin header should use the secure "https" URI scheme as well. 

See [RFC 6454 Section 3.1.1](https://datatracker.ietf.org/doc/html/rfc6454#section-3.1.1) for common Origin trust pitfalls. Two URIs on the same host, but containing "http" and "https" schemes are still considered as separate origins. 


I would expect something like the following snippet would be needed to resolve this issue:
```python
scheme = "https" if is_secure else "http"
headers.append("Origin: %s://%s" % (scheme, hostport)) 
```

**Comment by engn33r** [comment]
The RFC you referenced talks about same-origin policy. WebSockets don't use SOP. This is a known vulnerability: https://christian-schneider.net/CrossSiteWebSocketHijacking.html

Unless there is a strong reason to change the automated handling here, I am reluctant to automatically change the origin value from what the code currently does because it may in some cases break backwards compatibility because implementations must handle the origin header manually.

**Comment by jvpernis** [comment]
> Showing a https URL in the origin header can lead to a false sense of security because there is no SOP support in WebSockets.

I'm not sure that I understand your reasoning here. What sort of false sense of security would be implied by using a `https` URL? In that logic, there should be no Origin header automatically added at all, right?

When the handshake happens using a `https` URL, an Origin header containing `http` is just wrong, not somehow more secure.

To quote the blog you mentioned:

> Check the Origin header of the WebSocket handshake request on the server, since that header was designed to protect the server against attacker-initiated cross-site connections of victim browsers!

This is exactly where we are running into this bug. The connection handshake happens over `https`, which is why our server is configured to only accept the `https` in the Origin header.

As a workaround for this bug I have disabled checking the Origin header on our server altogether. I would rather lose one of our layers of protection than having to put up a big fat banner instructing other developers _not_ to remove the `http` origin since doing so would break compatibility with our Python client. 

Accepting the `http` origin is just too fragile. We have enough protection in other layers of our application to live without Origin checking for a while.

> [...] it may in some cases break backwards compatibility because implementations must handle the origin header manually

I agree that changing this behavior is considered a breaking change Fixing this therefore probably fits best in a new major release.


> Unless there is a strong reason to change the automated handling here, I am reluctant to automatically change the origin value from what the code currently does

That would be very unfortunate for us. While I respectfully disagree with your reasoning not to change the behavior, I want to thank you for all the hard work you have put in this project so far to make so many improvements!

**Comment by engn33r** [comment]
I reread this thread in full when I had more time on my hands, looked at what web browsers do when connecting to wss:// WebSockets, and determined that it could be beneficial to automatically set the Origin header to https:// when a handshake with a wss:// WebSocket is performed. The new version supporting this update is 1.3.0.

**Comment by snowman** [comment]
How to remove `Origin` header?

I can't connect to Chrome DevTools Protocol `webSocketDebuggerUrl` with python, while `nodejs` works

```sh
$ curl -v http://127.0.0.1:9222/json/list | jq 'map(select( .type == "page" and ( .url | contains("google.com") ) ))'
[
  {
    "description": "",
    "devtoolsFrontendUrl": "/devtools/inspector.html?ws=127.0.0.1:9222/devtools/page/B439C828C9AAA4B718B86EF5E7C0F28D",
    "faviconUrl": "https://www.google.com.hk/favicon.ico",
    "id": "B439C828C9AAA4B718B86EF5E7C0F28D",
    "parentId": "560704D038067627AEE1BA0971C018CF",
    "title": "Google",
    "type": "page",
    "url": "https://www.google.com",
    "webSocketDebuggerUrl": "ws://127.0.0.1:9222/devtools/page/B439C828C9AAA4B718B86EF5E7C0F28D"
  }
]
```

Figured out the differences of request headers of python and nodejs, python has extra `Origin` header,
and it shows error `websocket._exceptions.WebSocketBadStatusException: Handshake status 403 Forbidden`

**Comment by engn33r** [comment]
@snowman in the future, please make a new issue rather than posting to closed issues. The answer is in the docs https://websocket-client.readthedocs.io/en/latest/examples.html#suppress-origin-header
